src/api.ts

```diff
@@ -522,56 +577,153 @@ function createApi({
+      })
+
+      const styleToSave: StyleJSON = {
+        ...(await uncompositeStyle({
+          ...style,
+          sources,
+        })),
       }
 
-      db.prepare(
-        'INSERT INTO Style (id, stylejson) VALUES (:id, :stylejson)'
-      ).run({ id: styleId, stylejson: JSON.stringify(offlineStyle) })
+      db.prepare<{
+        id: string
+        stylejson: string
+        etag?: string
+        upstreamUrl?: string
+        sourceIdToTilesetId: string
+      }>(
+        'INSERT INTO Style (id, stylejson, etag, upstreamUrl, sourceIdToTilesetId) VALUES (:id, :stylejson, :etag, :upstreamUrl, :sourceIdToTilesetId)'
+      ).run({
+        id: styleId,
+        stylejson: JSON.stringify(styleToSave),
+        etag,
+        upstreamUrl,
+        sourceIdToTilesetId: JSON.stringify(sourceIdToTilesetId),
+      })
 
-      return addOfflineUrls(offlineStyle)
+      return {
+        ...addOfflineUrls({
+          sourceIdToTilesetId,
+          style: styleToSave,
+          styleId,
+        }),
+        id: styleId,
+      }
     },
 
-    async putStyle(id, style) {
-      db.prepare(
-        'INSERT INTO Style (id, stylejson) VALUES (:id, :stylejson)'
-      ).run({ id, stylejson: style })
-      return { ...style, id }
+    // TODO: May need to accept an access token
+    async updateStyle(id, style) {
+      if (!styleExists(id)) {
+        throw new NotFoundError(id)
+      }
+
+      const { sources, sourceIdToTilesetId } = await createOfflineSources({
+        sources: style.sources,
+      })
+
+      const styleToSave: StyleJSON = {
+        ...(await uncompositeStyle({
+          ...style,
+          sources,
+        })),
+      }
+
+      db.prepare<{
+        id: string
+        sourceIdToTilesetId: string
+        stylejson: string
+      }>(
+        'UPDATE Style SET stylejson = :stylejson, sourceIdToTilesetId = :sourceIdToTilesetId WHERE id = :id'
+      ).run({
+        id,
+        sourceIdToTilesetId: JSON.stringify(sourceIdToTilesetId),
+        stylejson: JSON.stringify(styleToSave),
+      })
+
+      return addOfflineUrls({
+        sourceIdToTilesetId,
+        style: styleToSave,
+        styleId: id,
+      })
     },
 
-    async listStyles(limit?: number) {
-      const baseQuery = 'SELECT stylejson FROM Style'
-      const stmt =
-        limit !== undefined
-          ? db.prepare(`${baseQuery} LIMIT ?`).bind(limit)
-          : db.prepare(baseQuery)
+    async listStyles() {
+      const styles: ({ name?: string } & IdResource)[] = []
 
-      return stmt
+      db.prepare('SELECT id, stylejson FROM Style')
         .all()
-        .map((row: { stylejson: string }) => JSON.parse(row.stylejson))
+        .forEach(({ id, stylejson }: { id: string; stylejson: string }) => {
+          try {
+            const style: StyleJSON = JSON.parse(stylejson)
+            // TODO: Should we have a fallback name here or let client handle?
+            styles.push({ id, name: style.name })
```

Comment:
Couple of questions here:

- Based on the style spec, `name` property of a style is optional. How should we handle this? Some potential options:
  - Client-side:
    - by file
      - require the user to provide a name to use, defaulting to some generic fallback name 
  - Server-side:
    - use some generic fallback name (since names don't have to be unique)
    
- For the initial iteration of this server, it might be more helpful for the client if we provide the entire stylejson definition here, since they can directly use it to render the list of map previews, as opposed to having to make an additional request for each style to get its stylejson. This would be reverted once proper thumbnail generation is supported